### PR TITLE
Add Romanian stopwords with s&t with comma

### PR DESCRIPTION
**Pull request by Trey314159**
Romanian uses s&t with commas (ș/ț), but for a long time those weren't available, so s&t with cedilla (ş/ţ) were used. Both are still in use, but the comma forms are much more common now. Both should be supported in stopword lists.

### CODE REVIEW

**Comment by rmuir** [comment]
Note, if we fix it here, stemmer maybe should deal with this case too? https://github.com/snowballstem/snowball/blob/master/algorithms/romanian.sbl#L26-L27

Alternatively, tokenfilter could be added that "normalizes/folds" these and runs before stopfilter and stemfilter to take care of it. It would have the advantage of giving the user choice (they can just create customanalzer and remove the normalization if they don't want that folding). But it would be overkill, if we should really just fix stopwords and stemmer.  Sorry, I'm not knowledgeable on Romanian, so I don't know if it causes problems to treat them "the same".

**Comment by Trey314159** [comment]
_Good catch!_ I didn't consider that the stemmer might also be of a vintage to only use the older orthography. I've contacted the Snowball mailing list (message not yet accepted) to suggest making the stemmer aware of both variants of each character. A character filter or token filter to convert the comma forms to cedilla forms seems like a reasonable hack for now, and I'll add that to our local implementation right away. I have to admit that it chafes a little to convert everything to the "wrong" form, but the internal representation is just an internal representation, I guess, as long as everything is consistent about it.

Feel free to decline this pull request if that makes more sense to you, though I think it might be a good bit of future proofing for each stage of the process to handle both forms, so as not to be reliant on a counterintuitive character mapping.

**Comment by rmuir** [comment]
I think we can merge this stopword list change anyway. But I think a filter may be worthwhile as a separate PR?

It has the advantage of making the terms conflate regardless of which variant of the character is being used. After reading up on the history of these characters, I think we should treat them "the same" for Romanian always. It will allow queries/documents to match in all cases. So I don't think of it as a hack, as the "different characters" may impact words outside of just stopwords and suffixes that are stemmed, too.

I'd recommend doing it as a TokenFilter. Many lucene analyzers for other languages have such normalizers as tokenfilters to deal with similar issues. CharFilter is not needed in this case as it won't impact tokenization, since StandardTokenizer etc will tokenize them all the same way (same unicode wordbreak properties).

**Comment by rmuir** [comment]
> I've contacted the Snowball mailing list (message not yet accepted)

fwiw I'm subscribed that list and haven't seen a message in 10 years. I think they are just using github issues/PRs now?

**Comment by Trey314159** [comment]
> After reading up on the history of these characters, I think we should treat them "the same" for Romanian always.

Yeah, I agree.

> I think a filter may be worthwhile as a separate PR?

Sounds good. A token filter called `romanian_normalization` or some such sounds good and in keeping with other analyzers.

> fwiw I'm subscribed that list and haven't seen a message in 10 years. I think they are just using github issues/PRs now?

It's definitely low volume. The archives show one message every month or two. If the moderator doesn't approve it by Monday I'll try something else. Please feel free to contact them through another channel if you can/want to.

> So I think the most ideal situation would be to both fix snowball and then map cedilla to "correct" forms with a TokenFilter?

Yeah. I guess if Snowball moves relatively quickly it would make sense to wait for them to make their change so as not to need to have a mapping going one way (comma to cedilla), and then switching it to be the opposite a short while later—though that _is_ what we're going to do locally since we aren't planning on upgrading particularly soon, while config changes are relatively easy.

**Comment by rmuir** [comment]
I think its actually quite easy to fix the stemmer if we want to just send them a pull request. I can help if you don't want to do it, I don't want to steal your thunder though :)

Thanks again for bringing this to our attention. It really looks like a serious issue. I downloaded a dump of Romanian wikipedia (100K+ articles) and generated some stats with https://scripts.sil.org/cms/scripts/page.php?site_id=nrsi&id=UnicodeCharacterCount:

```
proper comma forms
        U+0218  Ș       129164
        U+0219  ș       1602600
        U+021A  Ț       21578
        U+021B  ț       1088506

old cedilla forms
        U+015E  Ş       1007
        U+015F  ş       34008
        U+0162  Ţ       465
        U+0163  ţ       52129
```

Currently "old cedilla forms" are the only ones working correctly with lucene stemmer/stopwords :(

**Comment by Trey314159** [comment]
I looked into fixing the stemmer, and it got a little complicated, so if you want to have a go at it, please do. Otherwise we can try opening an issue on their repo. One problem I ran into is that they are building an ISO 8859-2 version of the stemmer, and the comma-based characters are not available. I'm not sure what the right path is, dropping ISO 8859-2 support, forking the stemmer into ISO 8859-2 and Unicode versions, or something else entirely.

As for your stats, that jives with what I found: ~350x more _tokens_ of "și" ("and") with comma than "şi" with cedilla, and ~50x more _articles_ with the comma variant. The tide has definitely shifted in favor of the comma variant, at least on Romanian Wikipedia.

BTW, if anyone needs example words to test the stemmer, here are a pair for each s and t, where the comma/cedilla character needs to be recognized properly to be stemmed properly: prelungeşte/prelungește and apreciaţi/apreciați.

**Comment by rmuir** [comment]
I took a stab. As far as the confusion there, IMO the direction is clear, to go forwards with Unicode. This is why Unicode exists :) Let's see what happens.

https://github.com/snowballstem/snowball/pull/177
https://github.com/snowballstem/snowball-data/pull/23

**Comment by strainu** [comment]
> Thanks again for bringing this to our attention. It really looks like a serious issue. I downloaded a dump of Romanian wikipedia (100K+ articles) and generated some stats with https://scripts.sil.org/cms/scripts/page.php?site_id=nrsi&id=UnicodeCharacterCount:
> 
> ```
> proper comma forms
>         U+0218  Ș       129164
>         U+0219  ș       1602600
>         U+021A  Ț       21578
>         U+021B  ț       1088506
> 
> old cedilla forms
>         U+015E  Ş       1007
>         U+015F  ş       34008
>         U+0162  Ţ       465
>         U+0163  ţ       52129
> ```
> 

For corectness' sake, I'd like to point out 2 things:
* the s-cedilla counts probably also include the Turkish names present in the Romanian Wikipedia, as well as file names (true for s-cedilla as well) 
* the site has scripts that actively convert to the correct diacritics, so the numbers are probably somewhat more optimistic than the reality.


> IMO the direction is clear, to go forwards with Unicode. This is why Unicode exists :) 

I fully agree.

**Comment by rmuir** [comment]
> For corectness' sake, I'd like to point out 2 things:
> 
>     * the s-cedilla counts probably also include the Turkish names present in the Romanian Wikipedia, as well as file names (true for s-cedilla as well)
> 
>     * the site has scripts that actively convert to the correct diacritics, so the numbers are probably somewhat more optimistic than the reality.

Thank you for this info. IMO the fact that characters in use are still inconsistent is a good argument for normalization. This is the approach that I attempted with the pull request to snowball.

A quick test of `site:ro` query from google, + sampling each result shows a lot more inconsistency. It still leans towards the s-comma/t-comma, though. So I think we are doing the right thing.

**Comment by rmuir** [comment]
The PR is not stale, see #13209 . I've been waiting on an official release to play it safe, but it has been too long: I think we should move forward here? I'd like to merge this after #13209

File: lucene/analysis/common/src/resources/org/apache/lucene/analysis/ro/stopwords.txt [hunk 1]
```diff
@@ -190,27 +207,34 @@ sale
 sau
 său
 se
+și
 şi
 sînt
 sîntem
+sînteți
```
Comment (stefanvodita) [inline lucene/analysis/common/src/resources/org/apache/lucene/analysis/ro/stopwords.txt:214]: If we have `sînt`/`sîntem`/`sînteți` shouldn't we have `sânt`/`sântem`/`sânteți` as well? Maybe that's a separate PR.

Comment (strainu) [reply lucene/analysis/common/src/resources/org/apache/lucene/analysis/ro/stopwords.txt:214]: No, `sânt` & co are not correct forms, the post-1993 forms are `sunt`/`suntem`/`sunteți`. See https://dexonline.ro/definitie/s%C3%A2ntem vs https://dexonline.ro/definitie/suntem

Comment (stefanvodita) [reply lucene/analysis/common/src/resources/org/apache/lucene/analysis/ro/stopwords.txt:214]: I think your point is the `î` spellings for "a fi" were used officially at some point, while the `â` spellings never were. It's tricky to talk about what is correct in a language though. Have people written text using the `â` spellings? Of course, switching `î` and `â` is a common pattern, which is "correct" for other words. Should we handle stemming for the `â` spellings then? I don't know, maybe it depends on how frequent the `â` spelling is in practice. I don't feel strongly about this one way or the other.

Comment (strainu) [reply lucene/analysis/common/src/resources/org/apache/lucene/analysis/ro/stopwords.txt:214]: I don't know enough about how Lucene is used at large to understand whether there would be a difference between a search for `noi suntem romani` vs `noi suntem români` vs `noi santem romani` vs `noi sântem romani` vs `noi suntem români` (`romani` means `Romans`, `români` means `Romanians`). I would recommend we leave this for another PR though, as the current one only tries to solve the technical problems related to S comma-below vs cedilla-below.

**Review by uschindler — APPROVED** [review]

**Comment by rmuir** [comment]
Thank you again @Trey314159 for raising this issue. I want to apologize again for it taking so long...

Stemmer and stopwords are now fixed, to complete the loop, I've opened #13233 to add a normalization filter to also take care of this issue for all the corner cases

**Comment by Trey314159** [comment]
Thanks, @rmuir for shepherding this through to completion!

**Pull request merged**